Fixed KMeans bug

NickHager opened the pull request:
Empty clustered centroids were sneaking through to the final set, which was causing an error when calculating and setting which class that related centroid represents. Centroids are removed in the final set if they have empty clusters.

NickHager committed:
Fixed KMeans bug

Empty clustered centroids were sneaking through to the final set, which was causing an error when calculating and setting which class that related centroid represents. Centroids are removed in the final set if they have empty clusters.

Rusty89 committed:
Added printing

Added printing, changed the tests to run the full set through k means and PAM with the number of clusters equal to the size of the edited set

Rusty89 commented:
Added some more stuff while we were at it

Rusty89 commented:
Also we need a hard limit on the amount of times K means will run.  With the now vastly increased number of clusters we are using, it will take a very very long time unless we limit it.

Rusty89 committed:
Fixed K means not terminating

K means will get rid of null clusters

Rusty89 committed:
Added some null checking to PAM

added some null checking to PAM and the distance metrics to prevent NaN and infinite results hopefully, I think this was what was causing null pointer exceptions in PAM

NickHager committed:
small correction to if statement

changed to 'continue'

Rusty89 reviewed (approved):
Rusty Clayton approves of this merge

Rusty89 committed:
Try/catch to stop PAM erroring out

Hopefully this works

Rusty89 committed:
Data set fixes

Wine data had an empty line at the end, the source of all our problems with PAM, quite a few of them had this actually

Rusty89 committed:
Added our own cluster choices back in

Giving us the option to just choose num clusters again

Rusty89 committed:
This is how clusters should be determined

Clusters size determined by size of edited sets

RialJohnson commented on Project 2/k-NNAlgorithms/src/Algorithms.java:
```diff
@@ -276,8 +277,29 @@
+            for (int i = 0; i < numClasses; i++){
+                if (clusters.get(i).size() < 1){
+                    clusters.remove(i);
+                    clusterCentroids.remove(i);
```
I want to refactor these to be while loops if we have time

RialJohnson commented on Project 2/k-NNAlgorithms/src/Data.java:
```diff
@@ -243,36 +272,35 @@ public void runKMeans(boolean regression, boolean euclidean, int numClusters){
             double RMSE = 0; // root mean squared error
             double absError = 0;
             for (int i = 0; i < numTrainingSets; i++) {
+                int numClusters = dataSets.trainingSets.size() / 4;
```
Let's comment why we are dividing by four here
